source/utils/HistoSampler.h

```diff
@@ -0,0 +1,51 @@
+#ifndef HISTO_SAMPLER_H
+#define HISTO_SAMPLER_H
+#include <vector>
+#include <G4RunManager.hh>
+#include <g4csv.hh>
+
+class HistoSampler1D{
+  unsigned int nbins    = 0;
+  unsigned int nentries = 0;
+  unsigned int nbins_x  = 0;
+  const tools::histo::h1d * h1G4;
+  std::vector<double> cdf;
+  std::vector<double> edges_x;
+  void create_cdf();
+public:
+  HistoSampler1D(const tools::histo::h1d * hG4):h1G4(hG4){};
```

Comment:
> As for this particular topic (in-class vs out-of-class inlined member functions), a quick (non-exhaustive, flawed, unreliable) search came up with one pronouncement against this style:
> 
> https://kks32.github.io/cpp-style-guide/?showone=Namespaces#Inline_Functions
> 
> For example, accessors and mutators should certainly be inside a class definition.
> 
> and zero in its favour.

Just for the sake of the argument: we have always tried to follow the advice at isocpp.org. This is what they say about inline definitions: https://isocpp.org/wiki/faq/inline-functions#inline-member-fns-more

Reply:
The argument seems to be

> it is harder on all the readers since it mixes what a class does (the external behavior) with how it does it (the implementation).

This sounds reasonable until you illustrate it with an example. The claim is that 

``` c++
namespace nexus {
  class Foo
  {
  public:
    Foo(Bartype b);
    Bartype GetBar() const;
    void SetBar(Bartype b);
  private:
    Bartype bar_;
  };
}
```

is clear to all readers, while extending it to

``` c++
namespace nexus {
  class Foo
  {
  public:
    Foo(Bartype bar) : bar_(bar) {}
    Bartype GetBar() const { return bar_ }
    void SetBar(Bartype bar) { bar_ = bar; }
  private:
    Bartype bar_;
  };
}
```

makes it sufficiently 'harder on all the readers' for it to be worth talking about.

Furthermore, the claim is that adding all the following junk to the bottom of the former (with its concomitant burden of DRY-violation), is *better* than turning the former into the latter.

``` c++
inline nexus::Foo::Foo(Bartype bar) : bar_(bar)
{}

inline Bartype nexus::Foo::GetBar() const
{ return bar_; }

inline void nexus::Foo::SetBar(Bartype bar)
{ bar_ = bar; }
```

Really?

Perhaps a more realistic example would be a typical HEP class with some 39 member functions, approximately 12 of which will be trivial. The claim then becomes that annotating (with half-line-long definitions) 12 of the 39 member function declarations in the class definition, hampers the understanding of the class structure.

Please understand, I spend a loooooot of my time thinking about clarity of code, and making code as legible as possible. I find this argument completely spurious. Why? Because we are talking about functions with *trivial* implementations, and those trivial implementations do not add a significant burden to the understanding of the declarations. On the other hand, there is the cost of DRY-violation. Furthermore it is debatable whether the clarity gained outweighs the burden of having to look elsewhere when you are interested in the implementation.

However, if you are convinced by or attached to this style, let's just stick to it. We have better things to do with our resources than debate this minor point.

Reply:
No, I'm not convinced or attached to that style. I was simply bringing to the discussion the reference I had followed historically. I'm not questioning your expertise and I appreciate the advice, but, in general, I like to understand the reasons behind guidelines. Otherwise, they just become _arguments from authority_.

Reply:
> I like to understand the reasons behind guidelines. Otherwise, they just become *arguments from authority*.

A highly commendable philosophy to which I subscribe.

> I'm not questioning your expertise

Why not? You should! I don't mean randomly or just for the sake of being contrarian, but you absolutely *should* question *why* I believe something, and I should be able to give you a reasonable answer. As I hope I have done in this case. And if you don't find my answer in this case convincing, then let's just stick to the old style, because this particular case is not important enough to merit any further discussion.